src/interface/control_textbox.cpp

```diff
@@ -288,70 +435,136 @@ ControlTextbox::on_key_down(const SDL_KeyboardEvent& key)
+    set_caret_pos(std::numeric_limits<int>::max());
+    set_secondary_caret_pos(0);
+    recenter_offset(true);
     return true;
   }
-  else if (key.keysym.sym == SDLK_RETURN)
+  else if (key.keysym.sym == SDLK_z && m_ctrl_pressed)
   {
-    parse_value();
+    if (m_undo_stack.empty())
+      return true;
+
+    const auto& change = m_undo_stack.back();
+    revert_value(change.string);
+    if (parse_value(true, TextAction::UNDO))
+    {
+      m_caret = std::move(change.carets.caret);
+      m_secondary_caret = std::move(change.carets.secondary_caret);
+    }
+    m_undo_stack.pop_back();
+    return true;
+  }
+  else if (key.keysym.sym == SDLK_y && m_ctrl_pressed)
+  {
+    if (m_redo_stack.empty())
+      return true;
+
+    const auto& change = m_redo_stack.back();
+    revert_value(change.string);
+    if (parse_value(true, TextAction::REDO))
+    {
+      m_caret = std::move(change.carets.caret);
+      m_secondary_caret = std::move(change.carets.secondary_caret);
+    }
+    m_redo_stack.pop_back();
     return true;
   }
 
   return false;
 }
 
 bool
-ControlTextbox::event(const SDL_Event& ev) {
+ControlTextbox::event(const SDL_Event& ev)
+{
   Widget::event(ev);
 
   if (ev.type == SDL_TEXTINPUT && m_has_focus)
+  {
     put_text(std::string(ev.text.text));
+    parse_value();
+  }
 
   return false;
 }
 
 bool
-ControlTextbox::parse_value(bool call_on_change /* = true  (see header)*/)
+ControlTextbox::parse_value(bool call_on_change /* = true  (see header)*/, TextAction action /* = TextAction::NONE  (see header)*/)
 {
-  // Abort if we have a validation function for the string, and the function
-  // says the string is invalid.
-  if (m_validate_string) {
-    if (!m_validate_string(this, get_contents())) {
-      revert_value();
-      return false;
-    }
+  const std::string new_str = get_contents();
+  if (m_internal_string_backup == new_str)
+    return false;
+
+  if (!validate_value())
+  {
+    revert_value(m_internal_string_backup);
+    m_caret = m_previous_caret_backup.caret;
+    m_secondary_caret = m_previous_caret_backup.secondary_caret;
+    return false;
+  }
+
+  if (action == TextAction::UNDO)
+  {
+    // Update redo stack
+    m_redo_stack.push_back({ m_internal_string_backup, std::move(m_previous_caret_backup) });
+  }
+  else
+  {
+    // Update undo stack
+    m_undo_stack.push_back({ m_internal_string_backup, std::move(m_previous_caret_backup) });
+    while (m_undo_stack.size() >= UNDO_STACK_SIZE)
+      m_undo_stack.erase(m_undo_stack.begin());
+
+    // Clear redo stack, if not currently performing a redo action
+    if (action != TextAction::REDO)
+      m_redo_stack.clear();
   }
 
-  std::string new_str = get_string();
-  if (m_internal_string_backup != new_str) {
-    m_internal_string_backup = new_str;
+  m_internal_string_backup = new_str;
+  m_previous_caret_backup = { m_caret, m_secondary_caret };
 
-    if (m_string)
-      *m_string = new_str;
+  if (m_string)
+    *m_string = new_str;
```

Comment:
what if instead of a pointer we just set some int variables or something, then returned a string view of some sort? Food for thought of the soul. It is certainly how most editors handle such logic, if I'm getting this down correctly